components/layout_2020/taffy/layout.rs

```diff
@@ -113,6 +115,19 @@ impl taffy::LayoutPartialTree for TaffyContainerContext<'_> {
         (*self.source_child_nodes[id]).borrow_mut().taffy_layout = *layout;
     }
 
+    #[allow(unsafe_code)]
+    fn resolve_calc_value(&self, val: *const (), basis: f32) -> f32 {
+        // SAFETY:
+        // - The calc `val` here is the same pointer we return to Taffy in `convert::length_percentage`
+        //   so it is safe to cast the type back to `*const CalcLengthPercentage`
+        // - Taffy guarantees that it never retains style values beyond the scope of it's style
+        //   computation methods, so we can be sure that the pointer we have passed it is still valid.
+        // - The reference we create here has a lifetime that does not escape this function, so it does
+        //   not matter if the pointer is later destroyed.
+        let calc = unsafe { &*(val as *const CalcLengthPercentage) };
+        calc.resolve(CSSPixelLength::new(basis)).px()
+    }
```

Comment:
Oof. I'm not a big fan of `unsafe` code in layout. Isn't there any way at all to avoid this?

Reply:
I have also tried to avoid `unsafe` in layout code. Taffy was `#[forbid(unsafe)]` prior to this feature. But I can't think
of a good way to do this without `unsafe`.

Reply:
FWIW, I don't think the `unsafe` here is a particularly scary one. The only invariants Servo needs to uphold are:

- That the pointer we return to Taffy is valid at the time it is returned
- That we cast the pointer back to the same type that we cast it from.
- That we don't store the pointer Taffy passes to `resolve_calc_value` anywhere.